src/arkode/arkode_arkstep.c

```diff
@@ -1729,23 +1740,88 @@ int arkStep_TakeStep_Z(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
     }
   }
 
-  /* Call the full RHS if needed e.g., an explicit first stage. If this is the
-     first step then we may need to evaluate or copy the RHS values from an
-     earlier evaluation (e.g., to compute h0). For subsequent steps treat this
-     RHS evaluation as an evaluation at the end of the just completed step to
-     potentially reuse (FSAL methods) or save (stiffly accurate methods with an
-     implicit first stage using Hermite interpolation) RHS evaluations from the
-     end of the last step. */
+  /* For a stiffly accurate implicit or ImEx method with an implicit first
+     stage, save f(tn, yn) if using Hermite interpolation as Fi[0] will be
+     overwritten during the implicit solve */
```

Comment:
The "complexity" of this function is growing out of control.  I'd like us as a group to think diligently about what we can do to simplify this.  Some options include:
* Create multiple `TakeStep` functions that are more specialized to specific purposes.
* Pull out distinct pieces of `TakeStep` into separate functions, each of which can encapsulate the necessary logic.  We could even have some ARKSTEP-specific internal function pointers, with separate implementations that handle simple-vs-complex situations.  For example, we could pull out everything that happens before we loop over stages into a `StepSetup` function, with separate implementations for ERK, DIRK, and ARK methods.
* Create more `sunbooleantype` flags that we determine during initialization, and that are used instead of re-determining so many things at each step.
* Deprecating some of our less-used options that necessitate so many corner cases.
* Reorganize ARKSTEP so that it is optimal for the default/simplest case (e.g., trivial predictor with autonomous RHS and Lagrange interpolant), and then do additional things only when doing something more complicated (e.g., higher-order predictors).

Reply:
> The "complexity" of this function is growing out of control. I'd like us as a group to think diligently about what we can do to simplify this. Some options include:
> 
>     * Create multiple `TakeStep` functions that are more specialized to specific purposes.

Having separate `TaksStep` functions for ERK and ARK/DIRK would help make the ERK case simpler. It's possible for a user to switch from an ERK to DIRK table during the integration so we'd need to update the `TakeStep` pointer when ever a method is attached.

>     * Pull out distinct pieces of `TakeStep` into separate functions, each of which can encapsulate the necessary logic.  We could even have some ARKSTEP-specific internal function pointers, with separate implementations that handle simple-vs-complex situations.  For example, we could pull out everything that happens before we loop over stages into a `StepSetup` function, with separate implementations for ERK, DIRK, and ARK methods.


I like the idea of having a `PreStep` function (and maybe a `PostStep` function too) to encapsulate this logic. 

>     * Create more `sunbooleantype` flags that we determine during initialization, and that are used instead of re-determining so many things at each step.

Many of these boolean values could be set a initialization but we'd need a flag to update these values any time a relevant option is changed e.g., if the method or predictor is changed during the integration. This could be forced with `ReInit` but that's does more than just validate the current settings.

>     * Deprecating some of our less-used options that necessitate so many corner cases.

I'm not sure I'd want to deprecate any of the options related to this logic since they could all reduce RHS evaluations. 

>     * Reorganize ARKSTEP so that it is optimal for the default/simplest case (e.g., trivial predictor with autonomous RHS and Lagrange interpolant), and then do additional things only when doing something more complicated (e.g., higher-order predictors).

I think the `PreStep` function is a better solution since the differences are before the actual step computations which are the same between the different configurations.